src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs

```diff
@@ -20,16 +23,31 @@ public static partial class SystemdHelpers
         /// </summary>
         /// <returns><c>True</c> if the current process is hosted as a systemd Service, otherwise <c>false</c>.</returns>
         public static bool IsSystemdService()
-            => _isSystemdService ?? (bool)(_isSystemdService = CheckParentIsSystemd());
+            => _isSystemdService ??= GetIsSystemdService();
 
-        private static bool CheckParentIsSystemd()
+        private static bool GetIsSystemdService()
         {
             // No point in testing anything unless it's Unix
             if (Environment.OSVersion.Platform != PlatformID.Unix)
             {
                 return false;
             }
 
+            // To support containerized systemd services, check if we're the main process (PID 1)
+            // and if there are systemd environment variables defined for notifying the service
+            // manager, or passing listen handles.
+#if NET5_0_OR_GREATER
+            int processId = Environment.ProcessId;
+#else
+            int processId = Process.GetCurrentProcess().Id;
+#endif
+
+            if (processId == 1)
```

Comment:
@eerhardt @judilsteve 

As I mentioned in the corresponding issue, this is a good way to check the current process is main process in the container.

In case we're not pid 1, we could extend it with checking if the pid 1 environment (`/proc/1/environ`) contains a `container` envvar. This file may be inaccessible. That is the way systemd itself checks if it runs as systemd in a container. And systemd sets this for its containerized services.

If we want to be looser, we could drop the checks and just rely on the envvars.

I'm fine with the implementation as is. I want to mention the options.